src/journal/test/JournalTest.hs

```diff
@@ -508,9 +516,138 @@ unit_bug14 = assertProgram ""
+
+prettyConcProgram :: ConcProgram -> String
+prettyConcProgram = show
+
+newtype History = History [Operation]
+  deriving Show
+
+newtype Pid = Pid Int
+  deriving (Eq, Ord, Show)
+
+data Operation
+  = Invoke Pid Command
+  | Ok     Pid Response
+  deriving Show
+
+toPid :: ThreadId -> Pid
+toPid tid = Pid (read (drop (length ("ThreadId " :: String)) (show tid)))
+
+prettyHistory :: History -> String
+prettyHistory = show
+
+concExec :: TQueue Operation -> Journal -> Command -> IO ()
+concExec queue jour cmd = do
+  pid <- toPid <$> myThreadId
+  atomically (writeTQueue queue (Invoke pid cmd))
+  -- Adds some entropy to the possible interleavings.
+  sleep <- randomRIO (5, 200)
+  threadDelay sleep
+  resp <- exec cmd jour
+  atomically (writeTQueue queue (Ok pid resp))
+
+-- Generate all possible single-threaded executions from the concurrent history.
+interleavings :: History -> Forest (Command, Response)
+interleavings (History [])  = []
+interleavings (History ops) =
+  [ Node (cmd, resp) (interleavings (History ops'))
+  | (tid, cmd)   <- takeInvocations ops
+  , (resp, ops') <- findResponse tid
+                      (filter1 (not . matchInvocation tid) ops)
+  ]
+  where
+    takeInvocations :: [Operation] -> [(Pid, Command)]
+    takeInvocations []                         = []
+    takeInvocations ((Invoke pid cmd)   : ops) = (pid, cmd) : takeInvocations ops
+    takeInvocations ((Ok    _pid _resp) : _)   = []
+
+    findResponse :: Pid -> [Operation] -> [(Response, [Operation])]
+    findResponse _pid []                                   = []
+    findResponse  pid ((Ok pid' resp) : ops) | pid == pid' = [(resp, ops)]
+    findResponse  pid (op             : ops)               =
+      [ (resp, op : ops') | (resp, ops') <- findResponse pid ops ]
+
+    matchInvocation :: Pid -> Operation -> Bool
+    matchInvocation pid (Invoke pid' _cmd) = pid == pid'
+    matchInvocation _   _                  = False
+
+    filter1 :: (a -> Bool) -> [a] -> [a]
+    filter1 _ []                   = []
+    filter1 p (x : xs) | p x       = x : filter1 p xs
+                       | otherwise = xs
+
+
+-- If any one of the single-threaded executions respects the state machine
+-- model, then the concurrent execution is correct.
+linearisable :: Forest (Command, Response) -> Bool
+linearisable = any' (go initModel)
+  where
+    go :: Model -> Tree (Command, Response) -> Bool
+    go model (Node (cmd, resp) ts) =
+      let
+        (model', resp') = step cmd model
+      in
+        resp == resp' && any' (go model') ts
+
+    any' :: (a -> Bool) -> [a] -> Bool
+    any' _p [] = True
+    any'  p xs = any p xs
+
+prop_concurrent :: Property
+prop_concurrent = forAllConcProgram $ \(ConcProgram cmdss) -> monadicIO $ do
+  -- Rerun a couple of times, to avoid being lucky with the interleavings.
+  replicateM_ 10 $ do
+    (fp, jour) <- run initJournal
+    queue <- run newTQueueIO
+    run (mapM_ (mapConcurrently (concExec queue jour)) cmdss)
```

Comment:
Hmm this seems suspicious. Shouldn't `mapM_` and `mapConcurrently` swap places? Or maybe I'm misunderstanding the meaning of ` newtype ConcProgram = ConcProgram [[Command]]` but I thought the outer list would be different (concurrent) programs and the inner list is just a sequential program.

Reply:
I did it like in qsm: each inner list is supposed to be executed concurrently, think of the outer list as barriers / synchronisation points between the concurrent execution.

Reply:
Hmm okay, must be me not remembering how we did it. I wonder if we could use these barriers in the linearisation check somehow..

Reply:
Perhaps...